File: go/objective/hash.go
package main

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"strings"
)

type Hasher struct {
	key byte
}

func NewHasher(key byte) *Hasher {
	return &Hasher{key: key}
}

func (h *Hasher) GenerateSalt(length int) (string, error) {
	bytes := make([]byte, length)
	if _, err := rand.Read(bytes); err != nil {
		return "", err
	}
	return hex.EncodeToString(bytes), nil
}

func (h *Hasher) ComputeHash(input string) string {
	var hash strings.Builder
	for _, b := range []byte(input) {
		hashedByte := b ^ h.key
		hash.WriteString(fmt.Sprintf("%02x", hashedByte))
	}
	return hash.String()
}

func main() {
	hasher := NewHasher(0xAA)
	salt, err := hasher.GenerateSalt(16)
	if err != nil {
		panic(err)
	}

	saltedInput := "password123" + salt
	hash := hasher.ComputeHash(saltedInput)

	fmt.Printf("Salt: %s\nHash: %s\n", salt, hash)
}

File: ocaml/objective/hash.ml
class hasher key =
  object (self)
    val mutable rand = Random.State.make_self_init ()

    method generate_salt ?(length=16) () =
      let salt = Buffer.create length in
      for _ = 1 to length do
        let byte = Random.State.int rand 256 in
        Buffer.add_string salt (Printf.sprintf "%02x" byte)
      done;
      Buffer.contents salt

    method xor_hash input =
      let buffer = Buffer.create (String.length input * 2) in
      String.iter (fun c ->
        let hashed_byte = Char.code c lxor key in
        Buffer.add_string buffer (Printf.sprintf "%02x" hashed_byte)
      ) input;  (* Fixed: Added closing parenthesis here *)
      Buffer.contents buffer
  end

let () =
  let hasher = new hasher 0xAA in
  let password = "password123" in
  let salt = hasher#generate_salt () in
  let salted_input = password ^ salt in
  let hash = hasher#xor_hash salted_input in
  Printf.printf "Salt: %s\nHash: %s\n" salt hash